mino/minows/rpc.go

```diff
@@ -0,0 +1,422 @@
+	}
+	return reply, nil
+}
+
+func (r rpc) openStream(ctx context.Context, dest address,
+	path string) (network.Stream, error) {
+	pid := protocol.ID(r.uri + path)
+	stream, err := r.mino.host.NewStream(ctx, dest.identity, pid)
+	if err != nil {
+		return nil, xerrors.Errorf("could not open stream: %v", err)
+	}
+
+	go func() {
+		<-ctx.Done()
+		err := stream.Reset()
+		if err != nil {
+			r.logger.Error().Err(err).Msg("could not reset stream")
+		}
+	}()
+
+	return stream, nil
+}
+
+func (r rpc) send(enc *gob.Encoder, msg serde.Message) error {
+	from := r.mino.myAddr.location.Bytes()
+
+	var payload []byte
+	if msg != nil {
+		bytes, err := msg.Serialize(r.context)
+		if err != nil {
+			return xerrors.Errorf("could not serialize message: %v", err)
+		}
+		payload = bytes
+	}
+
+	err := enc.Encode(&Packet{Source: from, Payload: payload})
+	if errors.Is(err, network.ErrReset) {
+		return err
+	}
+	if err != nil {
+		return xerrors.Errorf("could not encode packet: %v", err)
+	}
+	return nil
+}
+
+func (r rpc) receive(dec *gob.Decoder) (ma.Multiaddr, serde.Message, error) {
+	var packet Packet
+	err := dec.Decode(&packet)
+	if errors.Is(err, network.ErrReset) {
+		return nil, nil, err
+	}
+	if err != nil {
+		return nil, nil, xerrors.Errorf("could not decode packet: %v", err)
+	}
+
+	from, err := ma.NewMultiaddrBytes(packet.Source)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("could not unmarshal address: %v",
+			packet.Source)
+	}
+
+	if packet.Payload == nil {
+		return from, nil, nil
+	}
+	msg, err := r.factory.Deserialize(r.context, packet.Payload)
+	if err != nil {
+		return from, nil, xerrors.Errorf(
+			"could not deserialize message: %v",
+			err)
+	}
+	return from, msg, nil
+}
+
+func (r rpc) createOrchestrator(ctx context.Context,
+	initiator host.Host, streams []network.Stream) (*orchestrator, error) {
+	participant := r.mino.GetAddress().(address).location
+	myAddr, err := newOrchestratorAddr(participant, initiator.ID())
+	if err != nil {
+		return nil, xerrors.Errorf("could not create address: %v", err)
+	}
+
+	encoders := make(map[peer.ID]*gob.Encoder)
+	for _, stream := range streams {
+		encoders[stream.Conn().RemotePeer()] = gob.NewEncoder(stream)
+	}
+
+	o := &orchestrator{
+		logger: r.logger.With().Stringer("mino", myAddr).
+			Stringer("orchestrator", xid.New()).Logger(),
+		myAddr: myAddr,
+		rpc:    r,
+		outs:   encoders,
+		in:     make(chan Packet, MaxUnreadAllowed),
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(streams))
+	for _, stream := range streams {
+		go func(stream network.Stream) {
+			defer wg.Done()
+			decoder := gob.NewDecoder(stream)
+			for {
+				packet, err := o.listen(decoder)
+				if err != nil {
+					if strings.Contains(err.Error(), network.ErrReset.Error()) {
+						return
+					}
+					r.logger.Error().Err(err).Msg("message dropped")
+					continue
+				}
+				select {
+				case <-ctx.Done():
+					return
+				case o.in <- packet:
+				}
+			}
+		}(stream)
```

Comment:
Pre-Go 1.22, this could be outright buggy.
https://go.dev/wiki/CommonMistakes

Reply:
I don't see the problem here. The iterator is `stream`, which is passed as an argument to the anonymous go-function, which should be OK.

`o.in` is a variable from the function, and the channels should be mutex-safe.